### PR TITLE
update upstream

### PR DESCRIPTION
**Pull request by slwu89**
This PR does:

  1. various small updates to work with Julia >=1.8
  2. various small bugfixes
  3. adds examples/SIR-direct.jl

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1 +1,51 @@
-/docs/build/
+# Files generated by invoking Julia with --code-coverage
```
Comment (adolgert) [inline .gitignore:1]: You want an actual .gitignore? Great. It helps a lot to have a clear `git status` when it really is clear.

File: README.md [hunk 2]
```diff
@@ -7,3 +7,60 @@
 This is a continuous-time simulation with the simplest-possible state and non-exponential transitions. The goal is to make a compact, fastest-possible non-exponential continuous-time simulation.
 
 The simplest-possible state is a vector-addition system, which has a state that is a vector of integers just like chemical equations. A transition is defined by a vector of integers that represent how many values to take or give. However, for this simulation, the transition rates can be any function of the state and have exponential or non-exponential distributions.
+
+## Vector Addition System
+
+A [vector addition system](https://en.wikipedia.org/wiki/Vector_addition_system) (VAS) is a simple mathematical language to model systems. It is simpler than the Petri net language, but has the advantage of allowing for more free-form simulation design, as intensity functions can be arbitrary functions of state.
+
```
Comment (adolgert) [inline README.md:14]: This is a good start. The first paragraph is right for the README. Maybe the rest should be in the docs directory. It needs to be expanded, too. This is such a direct way to write a simulation, but I don't know people think of it often these days.

File: README.md [hunk 3]
```diff
@@ -7,3 +7,60 @@
...
+    VectorAdditionFSM: +VectorAdditionModel vam
+    VectorAdditionFSM: +Any sampler
+end
+
+VectorAdditionModel *-- VectorAdditionSystem
+VectorAdditionFSM *-- VectorAdditionModel
+```
+
+Simulation using the VAS requires concrete implementation of the interface:
+
```
Comment (adolgert) [inline README.md:39]: Yeah, this is not nearly enough to explain what those functions are. It does help me get an overview.

File: examples/SIR-direct.jl [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    event_cnt += 1
+end
+
+plot(
+    out[1:event_cnt, 1],
+    out[1:event_cnt, 2:4],
+    label=["S" "I" "R"],
+    xlabel="Time",
+    ylabel="Number"
+)
```
Comment (adolgert) [inline examples/SIR-direct.jl:60]: It says there's no newline at the end of the file.

File: examples/SIR-direct.jl [hunk 5]
```diff
@@ -0,0 +1,60 @@
+using Random
+using Plots
+using Distributions
+using Fleck
+
```
Comment (adolgert) [inline examples/SIR-direct.jl:5]: This wants a documentation comment at the top to say what it's an example of.

File: src/sample/direct.jl [hunk 6]
```diff
@@ -47,14 +49,14 @@ end
 
 function set_clock!(dc::DirectCall{T}, clock::T, distribution::Exponential,
         enabled::Symbol, rng::AbstractRNG) where {T}
-    if Base.:(==)(enabled, :Enabled)
+    if enabled == :Enabled
```
Comment (adolgert) [inline src/sample/direct.jl:52]: Ah! I was fighting with this in a previous Julia version. Nice that it's fixable!

File: src/vas.jl [hunk 7]
```diff
@@ -10,36 +13,71 @@ is an array of values. Negative numbers mean the transition needs to
 take this many tokens from the state, meaning the state at those indices
 must be an integer at least that large. Positive numbers mean the
 transition places tokens into the state. Unlike chemical simulations,
-the rate doesn't depend on the number of combinations of species present.
+the rate need not depend on the number of combinations of species present.
 """
-struct VectorAdditionSystem
+struct VectorAdditionSystem{T <: Function}
```
Comment (adolgert) [inline src/vas.jl:18]: Good, good, catch.

File: src/vas.jl [hunk 8]
```diff
@@ -10,36 +13,71 @@ is an array of values. Negative numbers mean the transition needs to
...
 
-function vas_input(vas::VectorAdditionSystem, transition_idx)
+Return a function taking a single argument `state` that applies the
+state change assocaited with the transition indexed by `transition_idx`.
+"""
+function vas_delta(vas::VectorAdditionSystem, transition_idx)
     state_change = vas.give[:, transition_idx] - vas.take[:, transition_idx]
     let delta = state_change
-        state -> begin state .+= state_change end
+        state -> begin state .+= delta end
```
Comment (adolgert) [inline src/vas.jl:54]: Again, this is right. Man.

File: test/runtests.jl [hunk 9]
```diff
@@ -2,9 +2,16 @@ using Fleck
 using SafeTestsets
 using Test
 
-@testset "Fleck.jl" begin
+@testset "test_prefixsearch.jl" begin
```
Comment (adolgert) [inline test/runtests.jl:5]: Ah, you want nested testsets. Sure. Sure.

**Review by adolgert — APPROVED** [review]
The only actual change I'd like is to move most of the explanation of the Vector Addition System to a document inside the docs/ directory. We can do that when the time comes, so I'll approve this and let's watch Github do its merging work.

File: test/runtests.jl (hunk 9 above)
Comment (slwu89) [reply test/runtests.jl:5]: I defer to your preferred convention. This is simply what I'm used to from collaborating with the AlgJulia folks.

File: src/vas.jl (hunk 7 above)
Comment (slwu89) [reply src/vas.jl:18]: Full specialization for increased memory locality!

File: src/sample/direct.jl (hunk 6 above)
Comment (slwu89) [reply src/sample/direct.jl:52]: Yes, apparently not a problem after >= 1.6

File: examples/SIR-direct.jl (hunk 5 above)
Comment (slwu89) [reply examples/SIR-direct.jl:5]: Long term what we can do is use Literate.jl and combine with Documenter.jl to build website docs from this. Or if someone write a Julia interface to Quarto that lets us directly build docs websites using Quarto that would be preferable. What [AlgebraicPetri.jl](https://github.com/AlgebraicJulia/AlgebraicPetri.jl) does is take the specially formatted .jl files in examples/ and then uses Literate.markdown to generate markdown docs in docs/build that Documenter.jl then weaves into the final docs webpage.

Comment (adolgert) [reply examples/SIR-direct.jl:5]: I haven't seen Literate.jl. I'm guessing it follows the ancient "literate programming" scheme. I'll take a look.

File: test/runtests.jl (hunk 9 above)
Comment (adolgert) [reply test/runtests.jl:5]: Once you're in the world of `@testset`, it doesn't matter whether it's nested. There is an advantage for localization of variables, but there is also a disadvantage for creating fixtures. Julia's testing environment is weak compared to Python's. I say we go with this suggestion. Always feels better to be modern.

Comment (slwu89) [reply test/runtests.jl:5]: They may be following the SciML style guide: https://github.com/SciML/SciMLStyle#tests-and-continuous-integration

File: examples/SIR-direct.jl (hunk 5 above)
Comment (slwu89) [reply examples/SIR-direct.jl:5]: I don't think we need to look now, and I can set it up when the time comes.

**Pull request merged**